File: goPoker/player.go
package main

import
(
	"fmt"
	"sort"
)

type Player struct {
	Hand      []Card
	HandRank  string
	NumRank   int
	HRank  []Card
}

func NewPlayer() Player {
	return Player{
		Hand:      []Card{},
		HandRank:  "",
		NumRank:   -1,
		HRank:  []Card{},
	}
}

func (p *Player) PushCard(c Card) {
	p.Hand = append(p.Hand, c)
}

func (p Player) ShowHand() {
	for _, c := range p.Hand {
		c.printCard()
		fmt.Print(" ")
	}
	if p.HandRank == "" {
		fmt.Println()
	} else {
		fmt.Println(" - " + p.HandRank)
	}
}

func (p *Player) RankHand() {
	sort.Slice(p.Hand, func(i, j int) bool {
		return p.Hand[i].Face < p.Hand[j].Face
	})
		if p.isFlush() && p.isRoyalStraight() {
			p.HandRank = "Royal Flush"
			p.NumRank = 0
			return
		}
	
		if p.isStraightFlush() {
			p.HandRank = "Straight Flush"
			p.NumRank = 1
			if p.Hand[0].Face == 1 {
				p.HRank = append(p.HRank, p.Hand[0])
			} else {
				p.HRank = append(p.HRank, p.Hand[4])
			}
			return
		}
	
		if p.isFourOfKind() {
			p.HandRank = "Four of a Kind"
			p.NumRank = 2
			i:=0
			for i<5{
				if(p.Hand[i].Face==1){
					p.Hand[i].Face=14
				}
				i++
			}
			sort.Slice(p.Hand, func(i, j int) bool {
				return p.Hand[i].Face < p.Hand[j].Face
			})

			p.HRank = append(p.HRank, p.Hand[1])
			return
		}
	
		if p.isFullHouse() {
			p.HandRank = "Full House"
			p.NumRank = 3
			i :=0
			for i<5{
				if(p.Hand[i].Face==1){
					p.Hand[i].Face=14
				}
				i++
			}
			sort.Slice(p.Hand, func(i, j int) bool {
				return p.Hand[i].Face < p.Hand[j].Face
			})
			p.HRank = append(p.HRank, p.Hand[2])
			return
		}
	
		if p.isFlush() {
			p.HandRank = "Flush"
			p.NumRank = 4
			return
		}
	
		if p.isStraight() {
			p.HandRank = "Straight"
			p.NumRank = 5
			if p.Hand[0].Face == 1 {
				p.HRank = append(p.HRank, p.Hand[0])
			} else {
				p.HRank = append(p.HRank, p.Hand[4])
			}
			return
		}
	
		if p.isThreeOfKind() {
			p.HandRank = "Three of a Kind"
			p.NumRank = 6
			
			p.HRank = append(p.HRank, p.Hand[2])
			return
		}
	
		if p.isTwoPair() {
			p.HandRank = "Two Pair"
			p.NumRank = 7
			i:=0
			for i<5{
				if(p.Hand[i].Face==1){
					p.Hand[i].Face=14
				}
				i++
			}
			cardCount := make(map[int]int)

			for _, card := range p.Hand {
				cardCount[card.Face]++
			}

			for _, card := range p.Hand {

				if cardCount[card.Face] >= 2 {
					p.HRank = append(p.HRank, card)
					cardCount[card.Face] = 0 // Avoid adding the same pair again
				}
			}
			for _, card := range p.Hand {
				
				if cardCount[card.Face] >= 1{
					p.HRank = append(p.HRank, card)
					cardCount[card.Face] = 0 // Avoid adding the same pair again
				}
			}
			return
		}
	
		if p.isPair() {
			p.HandRank = "Pair"
			p.NumRank = 8
			i:=0
			for i<5{
				if(p.Hand[i].Face==1){
					p.Hand[i].Face=14
				}
				i++
			}
			cardCount := make(map[int]int)

			for _, card := range p.Hand {
				cardCount[card.Face]++
			}

			for _, card := range p.Hand {
				if cardCount[card.Face] >= 2 {
					p.HRank = append(p.HRank, card)
					cardCount[card.Face] = 0 // Avoid adding the same pair again
				}
			}
			
			return
		}
	
		p.HandRank = "High Card"
		p.NumRank = 9
		if p.Hand[0].Face == 1 {
			p.Hand[0].Face=14
			sort.Slice(p.Hand, func(i, j int) bool {
				return p.Hand[i].Face < p.Hand[j].Face
			})
		}
		return
}
	
func (p Player) isFlush() bool {
	// Implementation for checking flush
	firstSuit := p.Hand[0].Suit
	for _, card := range p.Hand {
		if card.Suit != firstSuit{
			return false
		}
	}
	return true
}

func (p Player) isRoyalStraight() bool {
	// Implementation for checking royal flush
	if p.Hand[0].Face == 1 && p.Hand[1].Face == 10 && p.Hand[2].Face == 11 &&
		p.Hand[3].Face == 12 && p.Hand[4].Face == 13 {
		return true
	}
	return false
}

func (p Player) isStraightFlush() bool {
	if !p.isFlush() || !p.isStraight() {
		return false
	}

	return true
}

func (p Player) isFourOfKind() bool {
	countMap := make(map[int]int)

	for _, card := range p.Hand {
		countMap[card.Face]++
	}

	for _, count := range countMap {
		if count == 4 {
			return true
		}
	}

	return false
}

func (p Player) isFullHouse() bool {
	countMap := make(map[int]int)

	for _, card := range p.Hand {
		countMap[card.Face]++
	}

	hasThree := false
	hasTwo := false

	for _, count := range countMap {
		if count == 3 {
			hasThree = true
		} else if count == 2 {
			hasTwo = true
		}
	}

	return hasThree && hasTwo
}

func (p Player) isStraight() bool {
	uniqueFaces := make(map[int]bool)
	maxFace := 0
	minFace := 14 // Setting it to a value higher than the maximum possible face value

	for _, card := range p.Hand {
		if card.Face == 1 {
			uniqueFaces[14] = true // Ace can be treated as high or low, 14 or 1
		}
		uniqueFaces[card.Face] = true
		if card.Face > maxFace {
			maxFace = card.Face
		}
		if card.Face < minFace {
			minFace = card.Face
		}
	}

	// Check for sequential uniqueness of faces
	for i := minFace; i <= maxFace; i++ {
		if !uniqueFaces[i] {
			return false
		}
	}

	return true
}

func (p Player) isThreeOfKind() bool {
	countMap := make(map[int]int)

	for _, card := range p.Hand {
		countMap[card.Face]++
	}

	for _, count := range countMap {
		if count == 3 {
			return true
		}
	}

	return false
}

func (p Player) isTwoPair() bool {
	countMap := make(map[int]int)
	pairCount := 0

	for _, card := range p.Hand {
		countMap[card.Face]++
	}

	for _, count := range countMap {
		if count == 2 {
			pairCount++
			p.HRank=append(p.HRank,newCard(countMap[count],1))
		}
	}
	return pairCount == 2

}

func (p Player) isPair() bool {
	countMap := make(map[int]int)

	for _, card := range p.Hand {
		countMap[card.Face]++
	}

	for _, count := range countMap {
		if count == 2 {
			//p.HRank=append(p.HRank,newCard(countMap[count],1))
			//p.HRank[0].printCard()
			return true

		}
	}

	return false
}

File: rustyPoker/poker/src/player.rs

use crate::card::Card;
#[derive(Debug)]
#[derive(Clone)]
pub struct Player {
    pub hand_rank: String,
    pub num_rank: i32,
    pub hand: Vec<Card>,
    pub h_rank: Vec<Card>,
    //h: Vec<String>, // Assuming this was intended to store strings (not used in the provided code)
}


impl Player 
{
    // Constructor methods
    pub fn new() -> Self {
        Player {
            hand_rank: String::new(),
            num_rank: -1,
            hand: Vec::new(),
            h_rank: Vec::new(),
            //h: Vec::new(),
        }
    }

    pub fn push_card(&mut self, c: Card) {
        self.hand.push(c);
    }


    pub fn show_hand(&self) {
        for c in &self.hand {
            
            c.print_card();
            print!(" ")
            // Implement the Card print method here
            //println!("{:?}", c);
        }
        if self.hand_rank.is_empty() {
            println!();
        } else {
            println!(" - {}", self.hand_rank);
        }
    }

    pub fn with_cards(cards: [Card; 5]) -> Self {
        Player {
            hand_rank: String::new(),
            num_rank: -1,
            hand: cards.to_vec(),
            h_rank: Vec::new(),
           // h: Vec::new(),
        }
    }




    // Implement other methods like rank_hand, compare functions, and helpers...

    pub fn rank_hand(&mut self) {
        self.hand.sort_by_key(|card| card.face);

        if self.is_flush() && self.is_straight() {
            if self.hand[0].face == 1 && self.hand[4].face == 13 {
                self.hand_rank = String::from("Royal Flush");
                self.num_rank = 0;
                return;
            }
            self.hand_rank = String::from("Straight Flush");
            self.num_rank = 1;
            self.h_rank.push(if self.hand[0].face == 1 { self.hand[0].clone() } else { self.hand[4].clone() });
            return;
        }

        if self.is_four_of_kind() {
            self.hand_rank = String::from("Four of a Kind");
            self.num_rank = 2;
            self.h_rank.push(self.hand[1].clone()); // Must be one of the four
            return;
        }

        if self.is_full_house() {
            self.hand_rank = String::from("Full House");
            self.num_rank = 3;
            self.h_rank.push(self.hand[2].clone()); // Must be 3pair
            return;
        }

        if self.is_flush() {
            self.hand_rank = String::from("Flush");
            self.num_rank = 4;
            return;
        }

        if self.is_straight() {
            self.hand_rank = String::from("Straight");
            self.num_rank = 5;
            self.h_rank.push(self.hand[4].clone());
            return;
        }

        if self.is_three_of_kind() {
            self.hand_rank = String::from("Three of a Kind");
            self.num_rank = 6;
            self.h_rank.push(self.hand[2].clone());
            return;
        }

        if self.is_two_pair() {
            self.hand_rank = String::from("Two Pair");
            self.num_rank = 7;
            // No specific card to add to h_rank
            return;
        }

        if self.is_pair() {
            self.hand_rank = String::from("Pair");
            self.num_rank = 8;
            //self.h_rank.push(if self.hand[4].face == self.h_rank[0].face { self.hand[4].clone() } else { self.hand[2].clone() });
            return;
        }

        self.hand_rank = String::from("High Card");
        self.num_rank = 9;
    }


    fn is_straight(&self) -> bool {
        if self.hand[0].face == 1 && self.hand[1].face == 10 && self.hand[2].face == 11 && self.hand[3].face == 12 && self.hand[4].face == 13 {
            return true;
        }

        self.hand.windows(2).all(|pair| pair[1].face == pair[0].face + 1)
    }

    fn is_flush(&self) -> bool {
        let first_suit = self.hand[0].suit;
        self.hand.iter().all(|card| card.suit == first_suit)
    }


    fn is_four_of_kind(&self) -> bool {
        let mut face_counts = vec![0; 14]; // Index 0 represents Face 1 (Ace), and index 13 represents Face 13 (King)
        for card in &self.hand {
            face_counts[card.face as usize] += 1;
        }

        face_counts.iter().any(|&count| count == 4)
    }

    fn is_full_house(&self) -> bool {
        let mut face_counts = vec![0; 14];
        for card in &self.hand {
            face_counts[card.face as usize] += 1;
        }

        face_counts.iter().any(|&count| count == 3) && face_counts.iter().any(|&count| count == 2)
    }

    fn is_three_of_kind(&self) -> bool {
        let mut face_counts = vec![0; 14];
        for card in &self.hand {
            face_counts[card.face as usize] += 1;
        }

        face_counts.iter().any(|&count| count == 3)
    }

    fn is_two_pair(&self) -> bool {
        let mut face_counts = vec![0; 14];
        for card in &self.hand {
            face_counts[card.face as usize] += 1;
        }

        face_counts.iter().filter(|&&count| count == 2).count() == 2
    }

    fn is_pair(&self) -> bool {
        let mut face_counts = vec![0; 14];
        for card in &self.hand {
            face_counts[card.face as usize] += 1;
        }

        face_counts.iter().any(|&count| count == 2)
    }

}

